/* The Great Computer Language Shootout 
   http://shootout.alioth.debian.org/

   contributed by Isaac Gouy 
*/

using System;

   class Except
   {
      static int Lo = 0;
      static int Hi = 0;

      static void Main(string[] args){
         int n = 1;
         if (args.Length > 0) n = Int32.Parse(args[0]);
         for (int i = 0; i < n; i++) some_function(i);
         Console.WriteLine("Exceptions: HI={0} / LO={1}", Hi, Lo);
      }

      public static void some_function(int n){
         try {
            hi_function(n);
         } 
         catch (Exception e) {
            Console.WriteLine("We shouldn't get here: " + e.Message);
         }
      }

      public static void hi_function(int n){
         try {
            lo_function(n);
         } 
         catch (Hi_Exception e) {
            Hi++;
         }
      }


      public static void lo_function(int n){
         try {
            blowup(n);
         } 
         catch (Lo_Exception e){
            Lo++;
         }
      }


      public static void blowup(int n){
         if ((n % 2) == 0) throw new Lo_Exception();
         else throw new Hi_Exception();
      }
   }

   class Lo_Exception : System.Exception {
      public Lo_Exception() {}
   }

   class Hi_Exception : System.Exception {
      public Hi_Exception() {}
   }
